src/fast_layers.jl

```diff
@@ -24,84 +24,160 @@ initial_params(c::FastChain) = vcat(initial_params.(c.layers)...)
+in each pass which would otherwise slow down the computation, it defaults to false.
 
 Note that this function has specializations on `tanh` for a slightly faster
 adjoint with Zygote.
 """
-struct FastDense{F,F2} <: FastLayer
+struct FastDense{F,F2,C} <: FastLayer
   out::Int
   in::Int
   σ::F
   initial_params::F2
+  cache :: C
   bias::Bool
+  numcols::Int64
   function FastDense(in::Integer, out::Integer, σ = identity;
-                 bias = true, initW = Flux.glorot_uniform, initb = Flux.zeros32)
+                 bias = true, numcols=1, precache=false, initW = Flux.glorot_uniform, initb = Flux.zeros32)
     temp = ((bias == false) ? vcat(vec(initW(out, in))) : vcat(vec(initW(out, in)),initb(out)))
     initial_params() = temp
-    new{typeof(σ),typeof(initial_params)}(out,in,σ,initial_params,bias)
+    if precache == true
+      cache = (
+        W = zeros(out, in),
+        y = zeros(out, numcols),
+        r = zeros(out, numcols),
+        zbar = zeros(out, numcols),
+        Wbar = zeros(out, in),
+        xbar = zeros(in, numcols),
+        pbar = if bias == true
+            zeros((out*in)+out)
+          else
+            zeros(out*in)
+          end
+        )
+      else
+        cache = nothing
+      end
+      new{typeof(σ), typeof(initial_params), typeof(cache)}(out,in,σ,initial_params,cache,bias,numcols)
+      # new{typeof(σ),typeof(initial_params)}(out,in,σ,initial_params,bias)
   end
 end
 
 # (f::FastDense)(x,p) = f.σ.(reshape(uview(p,1:(f.out*f.in)),f.out,f.in)*x .+ uview(p,(f.out*f.in+1):lastindex(p)))
 (f::FastDense)(x,p) = ((f.bias == true) ? (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x .+ p[(f.out*f.in+1):end])) : (f.σ.(reshape(p[1:(f.out*f.in)],f.out,f.in)*x)))
 
 ZygoteRules.@adjoint function (f::FastDense)(x,p)
-  if !isgpu(p)
-    W = @view p[reshape(1:(f.out*f.in),f.out,f.in)]
-  else
-    W = reshape(@view(p[1:(f.out*f.in)]),f.out,f.in)
-  end
-
-  if f.bias == true
-    b = p[(f.out*f.in + 1):end]
-    r = W*x .+ b
-    ifgpufree(b)
+  if typeof(f.cache) <: Nothing
+    if !isgpu(p)
+      W = @view p[reshape(1:(f.out*f.in),f.out,f.in)]
+    else
+      W = reshape(@view(p[1:(f.out*f.in)]),f.out,f.in)
+    end
+    if f.bias == true
+      b = p[(f.out*f.in + 1):end]
+      r = W*x .+ b
+      ifgpufree(b)
+    else
+      r = W*x
+    end
+    y = f.σ.(r)
   else
-    r = W*x
-  end
-
-  # if typeof(x) <: AbstractVector
-  #   r = p[(f.out*f.in+1):end]
-  #   mul!(r,W,x,one(eltype(x)),one(eltype(x)))
-  # else
-  #   b = @view p[(f.out*f.in+1):end]
-  #   r = reshape(repeat(b,outer=size(x,2)),length(b),size(x,2))
-  #   mul!(r,W,x,one(eltype(x)),one(eltype(x)))
-  # end
-
-  y = f.σ.(r)
-
-  function FastDense_adjoint(ȳ)
-    if typeof(f.σ) <: typeof(tanh)
-      zbar = ȳ .* (1 .- y.^2)
-    elseif typeof(f.σ) <: typeof(identity)
-      zbar = ȳ
+    cols = length(size(x)) == 1 ? 1 : size(x)[2]
+    if !isgpu(p)
+      f.cache.W .= @view p[reshape(1:(f.out*f.in),f.out,f.in)]

```

Comment:
why is this cached?

Reply:
Did this just to avoid the pointer allocation from taking @view. Would you prefer it uncached?

Reply:
if uncached it causes further allocations when its transpose is taken [here](https://github.com/Abhishek-1Bhatt/DiffEqFlux.jl/blob/896fc48f9a82711a9ced560f5d543cbedfdcbfed/src/fast_layers.jl#L149)

Reply:
Interesting, view of a transpose allocates?